components/autocomplete/index.js

```diff
@@ -226,15 +226,39 @@ export class Autocomplete extends Component {
 	}
 
 	insertCompletion( range, replacement ) {
-		const container = document.createElement( 'div' );
-		container.innerHTML = renderToString( replacement );
-		while ( container.firstChild ) {
-			const child = container.firstChild;
-			container.removeChild( child );
-			range.insertNode( child );
-			range.setStartAfter( child );
-		}
+		// Wrap completions so we can treat them as tokens.
+		const tokenWrapper = document.createElement( 'span' );
+		tokenWrapper.classList.add( 'autocomplete-token' );
+
+		// Make tokens non-editable so they may be deleted but not modified.
+		tokenWrapper.contentEditable = false;
+
+		tokenWrapper.innerHTML = renderToString( replacement );
+
+		range.insertNode( tokenWrapper );
+		range.setStartAfter( tokenWrapper );
 		range.deleteContents();
+
+		/*
+		 * Add non-breaking space after a completion because:
```

Comment:
Doesn't this mean, though, that there will never be a break in between a mention and the next word? This could result in awkward-looking text flows.

For example, this mention at the end of the line causes `stuff @editor done` to go onto a new line:

<img width="773" alt="screen shot 2018-05-07 at 11 29 25" src="https://user-images.githubusercontent.com/612155/39681052-83d5cb5a-51ea-11e8-8eb1-448217afa53d.png">

Here's how it would look without non-breaking spaces:

<img width="757" alt="screen shot 2018-05-07 at 11 32 08" src="https://user-images.githubusercontent.com/612155/39681048-7f9c532e-51ea-11e8-9553-5149867937ea.png">

Maybe it's a compromise we're OK with. Just checking 🙂

Reply:
@noisysocks It is definitely a compromise. In my testing, I was unable to add a regular space and programmatically place the cursor afterward. This appears to be the issue or at least related:
https://bugs.chromium.org/p/chromium/issues/detail?id=428313

I thought about using zero-width non-breaking spaces as TinyMCE seems to do to move the cursor around [inline boundaries](https://codepen.io/tinymce/pen/EwLYBM), making it easier to specifically target and edit things like link text. But the character is not always removed by the editor, and it's invisible nature can cause problems. For example, in Firefox, when I backspace from one paragraph block into another with a trailing text node containing a zero-width non-breaking space, the cursor is not placed at the end of the text but rather one character before the end.

Inserting a visible non-breaking space is the best I've come up with to place the cursor and have editing work as expected after inserting an autocomplete token. I'm not totally OK with it but figured we could live with it.

I'd love to hear other ideas though.